torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -355,6 +355,101 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
   return {std::move(payload), std::move(tensors)};
 }
 
+TensorPipeEntry tensorpipeSerialize(const Message& rpcMessage) {
+  tensorpipe::Message tpMessage;
+  std::vector<torch::Tensor> reservedTensors;
+  std::vector<std::vector<uint8_t>> copiedTensors;
+
+  const std::vector<char>& payload = rpcMessage.payload();
+  c10::List<at::Tensor> tensors = cloneSparseTensors(rpcMessage.tensors());
+
+  // Payload
+  tpMessage.data = (uint8_t*)(payload.data());
+  tpMessage.length = payload.size();
+
+  // Metadata - encode rpc message type + message id
+  tpMessage.metadata = std::to_string(static_cast<int>(rpcMessage.type())) +
+      ":" + std::to_string(static_cast<int>(rpcMessage.id()));
+
+  // Tensors
+  tpMessage.tensors.reserve(tensors.size());
+  for (at::Tensor tensor : tensors) {
+    // Keep original user tensors and cloned sparse tensors
+    reservedTensors.push_back(tensor);
+    tensorpipe::Message::Tensor tpTensor;
+    torch::jit::Pickler pickler(
```

Comment:
Question, is there any reason for doing this for every individual tensor instead of using one pickler for the entirer tensors list? Is it because we cannot tell the boundary between tensor metas in the writer lambda?

Reply:
Iterating the tensors is mainly because of current tensorpipe::Message interface. We need to fill metadata separately to tensorpipe::Message::tensors according to Message interface. Or we can hack by putting all tensors' metadata into the first tensor metadata. As the first version of implementation, we may would like to keep the design clean.
It gets explained with more details and future interface change plan in the design quip doc, section Data Conversion